node/src/store.rs

```diff
@@ -1,14 +1,283 @@
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use async_trait::async_trait;
+use celestia_types::ExtendedHeader;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use tendermint::Hash;
+use thiserror::Error;
+use tracing::{info, instrument};
+
+type Result<T, E = StoreError> = std::result::Result<T, E>;
+
+#[async_trait]
+pub trait Store: Send + Sync + Debug {
+    async fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader>;
+    async fn get_by_height(&self, height: u64) -> Result<ExtendedHeader>;
+
+    async fn height(&self) -> u64;
+    async fn has(&self, hash: &Hash) -> bool;
+    async fn has_at(&self, height: u64) -> bool;
+
+    async fn append_single(&self, header: ExtendedHeader) -> Result<()>;
+
+    async fn append<I: IntoIterator<Item = ExtendedHeader>>(&self, headers: I) -> Result<()>
```

Comment:
`I` is already defined in the where-clause
```suggestion
    async fn append<I>(&self, headers: I) -> Result<()>
```

Reply:
removed, didn't notice that, thanks. 👍  
Weird, I'd expect clippy to catch me adding obviously redundant bounds